src/crypto/crypto_dh.cc

```diff
@@ -539,62 +539,70 @@ WebCryptoKeyExportStatus DHKeyExportTraits::DoExport(
-
-  EVPKeyCtxPointer ctx(EVP_PKEY_CTX_new(our_key.get(), nullptr));
+bool CreateStatelessDiffieHellmanContext(
+    const ManagedEVPPKey& our_key,
+    const ManagedEVPPKey& their_key,
+    char** buf,
+    size_t* out_size) {
+  const EVPKeyCtxPointer ctx(EVP_PKEY_CTX_new(our_key.get(), nullptr));
   if (!ctx ||
       EVP_PKEY_derive_init(ctx.get()) <= 0 ||
       EVP_PKEY_derive_set_peer(ctx.get(), their_key.get()) <= 0 ||
-      EVP_PKEY_derive(ctx.get(), nullptr, &out_size) <= 0)
-    return AllocatedBuffer();
+      EVP_PKEY_derive(ctx.get(), nullptr, out_size) <= 0)
+    return false;
 
-  AllocatedBuffer result = AllocatedBuffer::AllocateManaged(env, out_size);
-  CHECK_NOT_NULL(result.data());
+  *buf = MallocOpenSSL<char>(*out_size);
+  size_t remainder_size = *out_size;
+  if (EVP_PKEY_derive(ctx.get(),
+                      reinterpret_cast<unsigned char*>(*buf),
+                      &remainder_size) <= 0) {
+    OPENSSL_clear_free(buf, *out_size);
+    return false;
+  }
 
-  unsigned char* data = reinterpret_cast<unsigned char*>(result.data());
-  if (EVP_PKEY_derive(ctx.get(), data, &out_size) <= 0)
-    return AllocatedBuffer();
+  ZeroPadDiffieHellmanSecret(remainder_size, *buf, *out_size);
 
-  ZeroPadDiffieHellmanSecret(out_size, &result);
-  return result;
+  return true;
 }
 
-// The version of StatelessDiffieHellman that returns an AllocatedBuffer
-// is not threadsafe because of the AllocatedBuffer allocation of a
-// v8::BackingStore (it'll cause much crashing if we call it from a
-// libuv worker thread). This version allocates a ByteSource instead,
-// which we can convert into a v8::BackingStore later.
-// TODO(@jasnell): Eliminate the code duplication between these two
-// versions of the function.
-ByteSource StatelessDiffieHellmanThreadsafe(
+Local<Value> StatelessDiffieHellman(
     Environment* env,
-    ManagedEVPPKey our_key,
-    ManagedEVPPKey their_key) {
+    const ManagedEVPPKey& our_key,
+    const ManagedEVPPKey& their_key) {
+  char* buf = nullptr;
   size_t out_size;
-
-  EVPKeyCtxPointer ctx(EVP_PKEY_CTX_new(our_key.get(), nullptr));
-  if (!ctx ||
-      EVP_PKEY_derive_init(ctx.get()) <= 0 ||
-      EVP_PKEY_derive_set_peer(ctx.get(), their_key.get()) <= 0 ||
-      EVP_PKEY_derive(ctx.get(), nullptr, &out_size) <= 0)
-    return ByteSource();
-
-  char* buf = MallocOpenSSL<char>(out_size);
-  ByteSource out = ByteSource::Allocated(buf, out_size);
-
-  if (EVP_PKEY_derive(
-          ctx.get(),
-          reinterpret_cast<unsigned char*>(buf),
-          &out_size) <= 0) {
+  if (!CreateStatelessDiffieHellmanContext(our_key,
+                                           their_key,
+                                           &buf,
+                                           &out_size))
+    return Local<Value>();
+
+  return Buffer::New(env,
+                     buf,
+                     out_size,
+                     [](char* data, void* hint) {
+                       OPENSSL_free(data);
```

Comment:
@addaleax not quite sure how I can pass `out_size` to this so that I can call `OPENSSL_clear_free()`. Do u know of a way?

Reply:
@RaisinTen You can pass anything in the `hint`, generally speaking – I think that practically, passing `reinterpret_cast<void*>(static_cast<uintptr_t>(out_size))` as the hint and then reading it as `static_cast<size_t>(reinterpret_cast<uintptr_t>(hint))` in the callback should be fine, even though it’s always a bit sketchy to cast between pointers and integers like that :slightly_smiling_face:

Reply:
@addaleax ah nice, updated the PR to use `hint` to pass in `out_size`, PTAL